org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java

```diff
@@ -0,0 +1,208 @@
+
+		// Skip the leaf node, which is a dummy node
+		return leafSelectionRange.getParent();
+	}
+
+	private SelectionRange handleNodeSelectionRange(SelectionRange selectionRange, DOMNode node, int offset) {
+		switch (node.getNodeType()) {
+			case Node.ELEMENT_NODE:
+				return handleElementSelectionRange(selectionRange, (DOMElement) node, offset);
+			case Node.ATTRIBUTE_NODE:
+				return handleAttributeSelectionRange(selectionRange, (DOMAttr) node, offset);
+			case Node.DOCUMENT_TYPE_NODE:
+				return handleDoctypeSelectionRange(selectionRange, (DOMDocumentType) node, offset);
+			default:
+				return handleGenericNodeSelectionRange(selectionRange, node, offset);
+		}
+	}
+
+	private SelectionRange handleAttributeSelectionRange(SelectionRange selectionRange, DOMAttr attr, int offset) {
+		selectionRange = handleGenericNodeSelectionRange(selectionRange, attr, offset);
+		DOMElement ownerElement = attr.getOwnerElement();
+		if (ownerElement != null) {
+			selectionRange = handleGenericNodeSelectionRange(selectionRange, ownerElement, offset);
+		}
+		return selectionRange;
+	}
+
+	private SelectionRange handleElementSelectionRange(SelectionRange selectionRange, DOMElement element, int offset) {
+		XMLPositionUtility.selectStartTagName(element);
+		int contentStart = element.getStartTagCloseOffset() + 1;
+		int contentEnd = element.getEndTagOpenOffset();
+		if (contentStart != DOMNode.NULL_VALUE && contentEnd != DOMNode.NULL_VALUE && offset >= contentStart
+				&& offset <= contentEnd) {
+			Range range = XMLPositionUtility.createRange(contentStart, contentEnd, element.getOwnerDocument());
+			if (range != null && !range.equals(selectionRange.getRange())) {
+				SelectionRange parentSelectionRange = new SelectionRange();
+				parentSelectionRange.setRange(range);
+				selectionRange.setParent(parentSelectionRange);
+				selectionRange = parentSelectionRange;
+			}
+		} else if (contentStart != DOMNode.NULL_VALUE && offset < contentStart) {
+			SelectionRange parentSelectionRange = new SelectionRange();
+			parentSelectionRange.setRange(XMLPositionUtility.selectStartTagName(element));
+			selectionRange.setParent(parentSelectionRange);
+			selectionRange = parentSelectionRange;
+		} else if (offset <= element.getEndTagCloseOffset()) {
+			SelectionRange parentSelectionRange = new SelectionRange();
+			parentSelectionRange.setRange(XMLPositionUtility.selectEndTagName(element));
+			selectionRange.setParent(parentSelectionRange);
+			selectionRange = parentSelectionRange;
+		}
+		return handleGenericNodeSelectionRange(selectionRange, element, offset);
+	}
+
+	private SelectionRange handleDoctypeSelectionRange(SelectionRange selectionRange, DOMDocumentType doctype,
+			int offset) {
+		DOMRange internalSubset = doctype.getInternalSubsetNode();
+		if (internalSubset != null) {
+			int subsetStart = internalSubset.getStart();
+			int subsetEnd = internalSubset.getEnd();
+			// (subsetStart != -1 && subsetStart < offset && offset < subset)
```

Comment:
Stil needed ?

Reply:
This was just an explanation of why I didn't check `subsetEnd != -1`. I'll clean up the wording.